test/test_rpc.py

```diff
@@ -0,0 +1,180 @@
+            n = i + self.rank + 1
+            ret = dist.rpc_sync('worker%d' % dstRank, torch.add,
+                                args=(torch.ones(n, n), torch.ones(n, n)))
+            self.assertEqual(ret, torch.ones(n, n) * 2)
+
+    @_wrap_with_rpc
+    def test_py_built_in(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, min, args=(n, n + 1, n + 2))
+        self.assertEqual(ret, n)
+
+    @_wrap_with_rpc
+    def test_py_user_defined(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, my_function, args=(n, n + 1, n + 2))
+        self.assertEqual(ret, (n + n + 1 + n + 2))
+
+    @_wrap_with_rpc
+    def test_py_class_constructor(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, my_class, args=(n,))
+        self.assertEqual(ret.a, n)
+
+    @_wrap_with_rpc
+    def test_py_class_instance_method(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, my_class(2).my_instance_method, args=(n,))
+        self.assertEqual(ret, (2 + n))
+
+    @_wrap_with_rpc
+    def test_py_class_method(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, my_class.my_class_method, args=(n, n + 1))
+        self.assertEqual(ret, (n + n + 1))
+
+    @_wrap_with_rpc
+    def test_py_class_static_method(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, my_class.my_static_method, args=(n + 10,))
+        self.assertEqual(ret, (n + 10) > 10)
+
+    @_wrap_with_rpc
+    def test_py_multi_aync_call(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        fut1 = dist.rpc_async('worker%d' % dstRank, my_class.my_static_method, args=(n + 10,))
+        fut2 = dist.rpc_async('worker%d' % dstRank, min, args=(n, n + 1, n + 2))
+        self.assertEqual(fut1.wait(), (n + 10) > 10)
+        self.assertEqual(fut2.wait(), n)
+
+    @_wrap_with_rpc
+    def test_py_no_return_result(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, no_result)
+        self.assertEqual(ret, None)
+
+    @_wrap_with_rpc
+    def test_py_function_exception(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, no_result, args=(10,))
+        expected = "run_python_udf_internal caught exception: no_result() takes 0 positional arguments but 1 was given"
+        self.assertEqual(ret, expected)
+
+    @_wrap_with_rpc
+    def test_py_modify_global_var(self):
+        n = self.rank + 1
+        dstRank = n % self.world_size
+        ret = dist.rpc_sync('worker%d' % dstRank, modify_global_var, args=(True,))
```

Comment:
This could be flaky, in worst case, a worker could sent out it's request, but hasn't received a request yet. In that case, the global var it owns is not changed.

This is the code to simulate the worst case.
```Python
        dstRank = n % self.world_size
        import time; time.sleep(self.rank)
        ret = dist.rpc_sync('worker%d' % dstRank, modify_global_var, args=(True,))
        self.assertEqual(global_var, True)
```


Fix:
```Python
        dstRank = n % self.world_size
        ret = dist.rpc_sync('worker%d' % dstRank, modify_global_var, args=(True,))
        dist.barrier()
        self.assertEqual(global_var, True)
```

Reply:
good catch!

Reply:
Thinking more about this. It seems neither `dist.barrer()` nor the existing `sync_rpc` can guarantee a fix on this problem. Because they only enforces all sends are done, but there is no guarantee that all received messages are processed.

This can be fixed when we switch to `ThreadPool` for recv and let `sync_rpc` block until both send and recv are done. But the question is what should we wait for if received messages trigger subsequent sends (and those sends can trigger additional sends). 

I am thinking about the just doing following:
1. block send queue, and wait till all sends are done.
2. block recv queue, and wait till all recvs are done.

This means that we at least wait until existing sends are done and they are processed on the callee side (sufficient for this test). And we are not going to wait for additional send/recv triggered by existing sends. Does this make sense? @xush6528 @zhaojuanmao 

(this is not a change request for this PR, just discussion)

Reply:
In terms of changes, if we all agree that `dist.barrier` does not guarantee to fix the problem, shall we disable the test for now?

Reply:
sounds good, let me disable the test for now